### PR TITLE
Restore handling of physical processors in rankfiles

### PR DESCRIPTION
**Pull request by rhc54**
Restore handling of physical processors in rankfiles. Note that the prior implementation was likely incorrect as it falsely assumed that physical core indices were unique, which isn't always true. Stipulate that physical rankfiles can only include PU numbers, and bind the result to the core that contains that physical PU. Update the mpirun man page to cover the new use-case.

Cherry-pick of:
open-mpi/ompi@2a90788
open-mpi/ompi@d0704ef

@jsquyres I know you are thrilled, but I don't know anyone else that can handle it.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/19/

**Build Log**
_last 50 lines_

```
[...truncated 24536 lines...]
                          Specify base address for shared memory region
              MCA sshmem: parameter "sshmem_base_backing_file_dir" (current value: "/dev/shm", data source: default, level: 9 dev/all, type: string)
                          Specifies where backing files will be created when mmap is used and shmem_mmap_anonymous set to 0.
              MCA sshmem: parameter "sshmem_mmap_priority" (current value: "20", data source: default, level: 3 user/all, type: int)
                          Priority for sshmem mmap component (default: 20)
              MCA sshmem: parameter "sshmem_mmap_anonymous" (current value: "1", data source: default, level: 4 tuner/basic, type: int)
                          Select whether anonymous sshmem is used for mmap component (default: 1)
              MCA sshmem: parameter "sshmem_mmap_fixed" (current value: "1", data source: default, level: 4 tuner/basic, type: int)
                          Select whether fixed start address is used for shmem (default: 1)
              MCA sshmem: parameter "sshmem_sysv_priority" (current value: "30", data source: default, level: 3 user/all, type: int)
                          Priority for the sshmem sysv component (default: 30)
              MCA sshmem: parameter "sshmem_sysv_use_hp" (current value: "-1", data source: default, level: 4 tuner/basic, type: int)
                          Huge pages usage [0 - off, 1 - on, -1 - auto] (default: -1)
              MCA sshmem: parameter "sshmem_verbs_priority" (current value: "40", data source: default, level: 3 user/all, type: int)
                          Priority for sshmem verbs component (default: 40)
              MCA sshmem: parameter "sshmem_verbs_hca_name" (current value: "", data source: default, level: 3 user/all, type: string, synonyms: memheap_base_hca_name)
                          Preferred hca (default: the first)
              MCA sshmem: parameter "sshmem_verbs_mr_interleave_factor" (current value: "2", data source: default, level: 3 user/all, type: int, synonyms: memheap_base_mr_interleave_factor)
                          try to give at least N Gbytes spaces between mapped memheaps of other PEs that are local to me (default: 2)
              MCA sshmem: parameter "sshmem_verbs_shared_mr" (current value: "-1", data source: default, level: 4 tuner/basic, type: int)
                          Shared memory region usage [0 - off, 1 - on, -1 - auto] (default: -1)
+ local 'common_mca=--bind-to core -x SHMEM_SYMMETRIC_HEAP_SIZE=1024M'
+ local 'mca=--bind-to core -x SHMEM_SYMMETRIC_HEAP_SIZE=1024M'
+ timeout -s SIGKILL 3m oshrun -np 8 --bind-to core -x SHMEM_SYMMETRIC_HEAP_SIZE=1024M --mca spml yoda -mca pml ob1 -mca btl self,tcp /var/lib/jenkins/jobs/gh-ompi-release-v1.8-pr/workspace/ompi_install1/examples/oshmem_max_reduction
1/8 dst = 7 8 9
7/8 dst = 7 8 9
2/8 dst = 7 8 9
3/8 dst = 7 8 9
5/8 dst = 7 8 9
4/8 dst = 7 8 9
6/8 dst = 7 8 9
0/8 dst = 7 8 9
+ '[' no == no ']'
+ timeout -s SIGKILL 3m oshrun -np 8 --bind-to core -x SHMEM_SYMMETRIC_HEAP_SIZE=1024M --mca spml yoda -mca pml ob1 -mca btl self,sm /var/lib/jenkins/jobs/gh-ompi-release-v1.8-pr/workspace/ompi_install1/examples/oshmem_max_reduction
3/8 dst = 7 8 9
5/8 dst = 7 8 9
7/8 dst = 7 8 9
0/8 dst = 7 8 9
2/8 dst = 7 8 9
6/8 dst = 7 8 9
1/8 dst = 7 8 9
4/8 dst = 7 8 9
+ '[' yes == yes ']'
+ timeout -s SIGKILL 3m oshrun -np 8 --bind-to core -x SHMEM_SYMMETRIC_HEAP_SIZE=1024M --mca spml yoda -mca pml ob1 -mca btl self,vader /var/lib/jenkins/jobs/gh-ompi-release-v1.8-pr/workspace/ompi_install1/examples/oshmem_max_reduction
./hpc_tests/jenkins/ompi/ompi_jenkins.sh: line 115:  5528 Killed                  $timeout_exe oshrun -np $np $mca $spml_yoda -mca pml ob1 -mca btl self,vader ${exe_path} ${exe_args}
Build step 'Execute shell' marked build as failure
[BFA] Scanning build for known causes...

[BFA] Done. 0s
```

Test FAILed.

**Comment by mike-dubman** [comment]
retest this please

**Comment by mellanox-github** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/20/
Test PASSed.

File: opal/mca/hwloc/base/hwloc_base_util.c [hunk 1]
```diff
@@ -66,16 +68,38 @@ hwloc_obj_t opal_hwloc_base_get_pu(hwloc_topology_t topo,  int lid)
        So first we have to see if we can find *any* cores by looking
        for the 0th core.  If we find it, then try to find the Nth
        core.  Otherwise, try to find the Nth PU. */
-    if (NULL == hwloc_get_obj_by_type(topo, HWLOC_OBJ_CORE, 0)) {
+    if (NULL == (obj = hwloc_get_obj_by_type(topo, HWLOC_OBJ_CORE, 0))) {
```
Comment (jsquyres) [inline opal/mca/hwloc/base/hwloc_base_util.c]: This assignment to obj is never used.

File: orte/mca/rmaps/rank_file/rmaps_rank_file_component.c [hunk 2]
```diff
@@ -92,6 +92,14 @@ static int orte_rmaps_rank_file_register(void)
                                           MCA_BASE_VAR_SCOPE_READONLY, &orte_rankfile);
     (void) mca_base_var_register_synonym(tmp, "orte", "orte", NULL, "rankfile", 0);
 
+    mca_rmaps_rank_file_component.physical = false;
+    (void) mca_base_component_var_register(c, "physical", "Rankfile contains physical cpu designations",
+                                           MCA_BASE_VAR_TYPE_BOOL, NULL, 0, 0,
+                                           OPAL_INFO_LVL_9,
```
Comment (jsquyres) [inline orte/mca/rmaps/rank_file/rmaps_rank_file_component.c]: I'd make this LVL_5.

**Comment by jsquyres** [comment]
Aside from those minor quibbles, looks ok to me.

**Comment by rhc54** [comment]
Out of curiosity: do we have any evidence of users actually understanding and using the opal var levels? It seems that everyone on the user list always does "all" to get around the change we instituted as nobody can remember what these levels are supposed to mean, and assume that if a param doesn't show up, then it just doesn't exist (and not that it must be at a different level).

Just wondering. I can change these, but I'm wondering if we are wasting our time.

**Comment by rhc54** [comment]
Sigh - blasted git really hosed this branch. Will resubmit and try to fix.

**Comment by mellanox-github** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/43/
Test PASSed.

**Comment by goodell** [comment]
This was superseded by #103, right?

**Comment by rhc54** [comment]
Yes

**Pull request closed without merging**